src/input_map.rs

```diff
@@ -166,31 +164,39 @@ impl<A: Actionlike> InputMap<A> {
 impl<A: Actionlike> InputMap<A> {
     /// Insert a mapping between `action` and `input`
     ///
-    /// Existing mappings for that action will not be overwritten.
-    /// If the set for this action is already full, this insertion will silently fail.
+    /// # Panics
+    /// Panics if the map is full and `input` is not a duplicate or `input` is [`UserInput::Null`].
     pub fn insert(&mut self, action: A, input: impl Into<UserInput>) -> &mut Self {
         let input = input.into();
 
         // Don't insert Null inputs into the map
         if input == UserInput::Null {
-            return self;
+            panic!("Inserted input can't be {input:?}")
         }
 
-        // Don't overflow the set!
-        if self.n_registered(action.clone(), None) >= 16 {
-            return self;
-        }
+        self.map[action.index()].insert(input);
 
-        // Respect any per-input-mode caps that have been set
-        if let Some(per_mode_cap) = self.per_mode_cap {
-            for input_mode in input.input_modes() {
-                if self.n_registered(action.clone(), Some(input_mode)) >= per_mode_cap {
-                    return self;
-                }
-            }
+        // Cache clashes now, to ensure a clean state
+        self.cache_possible_clashes();
+
+        self
+    }
+
+    /// Insert a mapping between `action` and `input` at the provided index
+    ///
+    /// If a matching input already existed in the set, it will be moved to the supplied index. Any input that was previously there will be moved to the matching input’s original index.
+    ///
+    /// # Panics
+    /// Panics if the map is full and `input` is not a duplicate or `input` is [`UserInput::Null`].
+    pub fn insert_at(&mut self, action: A, input: impl Into<UserInput>, index: usize) -> &mut Self {
+        let input = input.into();
+
+        // Don't insert Null inputs into the map
+        if input == UserInput::Null {
```

Comment:
Can you please remove UserInput::Null in this PR? @sixfold correctly pointed out that this wasn't very useful.

Reply:
You mean to remote it at all?

Reply:
If yes, how would you like to handle chords coertion if there is no inputs?

Reply:
I meant remove it entirely, yes.

Which chord coercion in particular were you wondering about?

Reply:
I mean how to handle 0 in this case:

https://github.com/Leafwing-Studios/leafwing_input_manager/blob/2496c38d756275b7acedb19a5fc62b57a2632fbd/src/buttonlike_user_input.rs#L52-L56

Just create an empty chord?

Reply:
Yeah, I think we should create an n-length chord in all cases there, rather than trying to be clever.